error.c

```diff
@@ -1603,6 +1603,29 @@ exc_cause(VALUE exc)
     return rb_attr_get(exc, id_cause);
 }
 
+/*
+ * call-seq:
+ *   exception.root_cause   -> an_exception
+ *
+ * The innermost exception wrapped by this exception.
+ * This is useful for unwrapping exceptions and discovering their original exception.
+ */
+
+static VALUE
+exc_root_cause(VALUE exc)
+{
+    VALUE current = exc;
+
+    while (true) {
+        VALUE latest_cause = exc_cause(current);
+        if (NIL_P(latest_cause)) return current;
+
+        current = latest_cause;
+    }
+
+    UNREACHABLE;
```

Comment:
`UNREACHABLE` may be ineffective depending on compilers.
```suggestion
    UNREACHABLE_RETURN(Qnil);
```